charms/k8s/src/charm.py

```diff
@@ -97,64 +116,38 @@ def _get_snap_version(self) -> Optional[str]:
         if match:
             return match.group()
 
-        logger.info("Snap k8s not found or no version available.")
+        log.info("Snap k8s not found or no version available.")
         return None
 
+    @on_error(ops.BlockedStatus("Failed to install k8s snap."), SnapError)
     def _install_k8s_snap(self):
         """Install the k8s snap package."""
-        self.unit.status = ops.MaintenanceStatus("Installing k8s snap")
+        status.add(ops.MaintenanceStatus("Installing k8s snap"))
         k8s_snap = self.snap_cache["k8s"]
         if not k8s_snap.present:
             channel = self.config["channel"]
             k8s_snap.ensure(SnapState.Latest, channel=channel)
 
-    def _on_install(self, event):
-        """Handle install event for the charm.
-
-        Args:
-            event: The event that triggered this handler.
-        """
-        # TODO: Implement clustering using leader units.
-        try:
-            self._install_k8s_snap()
-            self._apply_snap_requirements()
-            self._bootstrap_k8s_snap()
-            self._enable_components()
-            self.unit.status = ops.WaitingStatus("Waiting for K8s to be ready")
-        except InvalidResponseError as e:
-            logger.warning("Failed to query k8s snap. Reason: %s", e)
-            self.unit.status = ops.WaitingStatus("Waiting for K8sd API.")
-            event.defer()
-        except SnapError as e:
-            logger.error("Failed to install k8s snap. Reason: %s", e.message)
-            self.unit.status = ops.BlockedStatus("Failed to install k8s snap")
-            event.defer()
-        except subprocess.CalledProcessError as e:
-            logger.error("Failed to run subprocess: %s", e)
-            event.defer()
-        except K8sdConnectionError as e:
-            logger.warning("Unable to contact K8sd API: %s", e)
-            self.unit.status = ops.WaitingStatus("Waiting for K8sd API.")
-            event.defer()
-
-    def _on_update_status(self, event: ops.UpdateStatusEvent):
-        """Handle update-status event.
-
-        Args:
-            event: event triggering the handler.
-        """
-        try:
-            if self.api_manager.is_cluster_ready():
-                if version := self._get_snap_version():
-                    self.unit.set_workload_version(version)
-                self.unit.status = ops.ActiveStatus("Ready")
-        except InvalidResponseError:
-            self.unit.status = ops.WaitingStatus("Waiting for K8s to be ready")
-            event.defer()
-        except K8sdConnectionError:
-            logger.exception("Unable to contact K8sdAPI")
-            self.unit.status = ops.WaitingStatus("Waiting for K8sd API")
-            event.defer()
+    @on_error(
+        ops.WaitingStatus("Failed to update status"),
```

Comment:
maybe something like this -- we're "waiting" because it could be ready later right?  And the user doesn't need to do anything?

```suggestion
        ops.WaitingStatus("Cluster not yet ready"),
```